## 1.0 Data Preparation & Data Cleaning

### Description:

The purpose of this step is to prepare and clean the data for the project. I will perform data cleaning tasks such as handling missing values, removing duplicates, and transforming data. Also, I will do feature engineering to improve the quality of the data and make it more suitable for analysis.

The column definitions and their roles in the dataset are summarized in the table below:

|  | Variable Name                      | Type | Category | Description |
| :--- |:-----------------------------------| :--- | :--- | :--- |
| 1 | **Marital Status**                 | Integer | Demographic | 1-single, 2-married, 3-widower, 4-divorced, 5-facto union, 6-legally separated. |
| 2 | **Application Mode**               | Integer | Academic | How the student applied (e.g., general, special contingent, transfer). |
| 3 | **Application Order**              | Integer | Academic | Preference (0=1st choice to 9=last choice). |
| 4 | **Course**                         | Integer | Academic | Degree program (e.g., Nursing, Informatics, Management). |
| 5 | **Daytime/Evening**                | Integer | Academic | 1=daytime, 0=evening. |
| 6 | **Previous Qualification**         | Integer | Academic | Education level before entry (e.g., Secondary, Bachelor's). |
| 7 | **Previous Qualification (Grade)** | Continuous | Academic | Grade of previous qualification (0-200). |
| 8 | **Nationality**                    | Integer | Demographic | Student's country of origin. |
| 9 | **Mother's Qualification**         | Integer | Socio-economic | Highest education level of the mother. |
| 10 | **Father's Qualification**         | Integer | Socio-economic | Highest education level of the father. |
| 11 | **Mother's Occupation**            | Integer | Socio-economic | Occupation category of the mother. |
| 12 | **Father's Occupation**            | Integer | Socio-economic | Occupation category of the father. |
| 13 | **Admission Grade**                | Continuous | Academic | Grade for university admission (0-200). |
| 14 | **Displaced**                      | Integer | Demographic | 1=yes, 0=no (lives away from home). |
| 15 | **Educational special needs**      | Integer | Demographic | 1=yes, 0=no (educational special needs). |
| 16 | **Debtor**                         | Integer | Socio-economic | 1=yes, 0=no (has outstanding debt). |
| 17 | **Tuition fees up to date**        | Integer | Socio-economic | 1=yes, 0=no (tuition fees paid). |
| 18 | **Gender**                         | Integer | Demographic | 1=Male, 0=Female. |
| 19 | **Scholarship holder**                   | Integer | Socio-economic | 1=yes, 0=no (scholarship holder). |
| 20 | **Age at Enrollment**              | Integer | Demographic | Student's age when starting. |
| 21 | **International**                  | Integer | Demographic | 1=yes, 0=no. |
| 22 | **Curricular units 1st sem (credited)**          | Integer | Performance | Number of units credited in 1st semester. |
| 23 | **Curricular units 1st sem (enrolled)**          | Integer | Performance | Number of units enrolled in 1st semester. |
| 24 | **Curricular units 1st sem (evaluations)**             | Integer | Performance | Number of units evaluated in 1st semester. |
| 25 | **Curricular units 1st sem (approved)**             | Integer | Performance | Number of units approved in 1st semester. |
| 26 | **Curricular units 1st sem (grade)**             | Continuous | Performance | Average grade in 1st semester (0-20). |
| 27 | **Curricular units 1st sem (without evaluations)**          | Integer | Performance | Number of units without evaluation in 1st semester. |
| 28 | **Curricular units 2nd sem (credited)**          | Integer | Performance | Number of units credited in 2nd semester. |
| 29 | **Curricular units 2nd sem (enrolled)**          | Integer | Performance | Number of units enrolled in 2nd semester. |
| 30 | **Curricular units 2nd sem (evaluations)**             | Integer | Performance | Number of units evaluated in 2nd semester. |
| 31 | **Curricular units 2nd sem (approved)**             | Integer | Performance | Number of units approved in 2nd semester. |
| 32 | **Curricular units 2nd sem (grade)**             | Continuous | Performance | Average grade in 2nd semester (0-20). |
| 33 | **Curricular units 2nd sem (without evaluations)**          | Integer | Performance | Number of units without evaluation in 2nd semester. |
| 34 | **Unemployment Rate**              | Continuous | Macro-econ | Regional unemployment rate (%). |
| 35 | **Inflation Rate**                 | Continuous | Macro-econ | Current inflation rate (%). |
| 36 | **GDP**                            | Continuous | Macro-econ | Gross Domestic Product (GDP). |
| 37 | **Target**                         | Categorical | Target | Student outcome: **Dropout**, **Enrolled**, or **Graduate**. |

### Detailed Categorical Mappings

To keep the table clean, the detailed codes for Application Mode, Course, Qualifications, and Occupations are maintained below for reference during data cleaning.

#### Application Mode
*   **1** - 1st phase - general contingent
*   **2** - Ordinance No. 612/93
*   **5** - 1st phase - special contingent (Azores Island)
*   **7** - Holders of other higher courses
*   **10** - Ordinance No. 854-B/99
*   **15** - International student (bachelor)
*   **16** - 1st phase - special contingent (Madeira Island)
*   **17** - 2nd phase - general contingent
*   **18** - 3rd phase - general contingent
*   **26** - Ordinance No. 533-A/99, item b2) (Different Plan)
*   **27** - Ordinance No. 533-A/99, item b3 (Other Institution)
*   **39** - Over 23 years old
*   **42** - Transfer
*   **43** - Change of course
*   **44** - Technological specialization diploma holders
*   **51** - Change of institution/course
*   **53** - Short cycle diploma holders
*   **57** - Change of institution/course (International)

#### Course
*   **33** - Biofuel Production Technologies
*   **171** - Animation and Multimedia Design
*   **8014** - Social Service (evening attendance)
*   **9003** - Agronomy
*   **9070** - Communication Design
*   **9085** - Veterinary Nursing
*   **9119** - Informatics Engineering
*   **9130** - Equinculture
*   **9147** - Management
*   **9238** - Social Service
*   **9254** - Tourism
*   **9500** - Nursing
*   **9556** - Oral Hygiene
*   **9670** - Advertising and Marketing Management
*   **9773** - Journalism and Communication
*   **9853** - Basic Education
*   **9991** - Management (evening attendance)

#### Qualifications (Mother, Father, and Previous)
*   **1** - Secondary Education - 12th Year of Schooling or Eq.
*   **2** - Higher Education - Bachelor's Degree
*   **3** - Higher Education - Degree
*   **4** - Higher Education - Master's
*   **5** - Higher Education - Doctorate
*   **6** - Frequency of Higher Education
*   **9** - 12th Year of Schooling - Not Completed
*   **10** - 11th Year of Schooling - Not Completed
*   **11** - 7th Year (Old)
*   **12** - Other - 11th Year of Schooling
*   **14** - 10th Year of Schooling
*   **18** - General commerce course
*   **19** - Basic Education 3rd Cycle (9th/10th/11th Year) or Equiv.
*   **22** - Technical-professional course
*   **26** - 7th year of schooling
*   **27** - 2nd cycle of the general high school course
*   **29** - 9th Year of Schooling - Not Completed
*   **30** - 8th year of schooling
*   **34** - Unknown
*   **35** - Can't read or write
*   **36** - Can read without having a 4th year of schooling
*   **37** - Basic education 1st cycle (4th/5th year) or equiv.
*   **38** - Basic Education 2nd Cycle (6th/7th/8th Year) or Equiv.
*   **39** - Technological specialization course
*   **40** - Higher education - degree (1st cycle)
*   **41** - Specialized higher studies course
*   **42** - Professional higher technical course
*   **43** - Higher Education - Master (2nd cycle)
*   **44** - Higher Education - Doctorate (3rd cycle)

#### Occupations (Mother and Father)
*   **0** - Student
*   **1** - Representatives of the Legislative Power and Executive Bodies, Directors, Directors and Executive Managers
*   **2** - Specialists in Intellectual and Scientific Activities
*   **3** - Intermediate Level Technicians and Professions
*   **4** - Administrative staff
*   **5** - Personal Services, Security and Safety Workers and Sellers
*   **6** - Farmers and Skilled Workers in Agriculture, Fisheries and Forestry
*   **7** - Skilled Workers in Industry, Construction and Craftsmen
*   **8** - Installation and Machine Operators and Assembly Workers
*   **9** - Unskilled Workers
*   **10** - Armed Forces Professions
*   **90** - Other Situation
*   **99** - (blank)
*   **122** - Health professionals
*   **123** - teachers
*   **125** - Specialists in information and communication technologies (ICT)
*   **131** - Intermediate level science and engineering technicians and professions
*   **132** - Technicians and professionals, of intermediate level of health
*   **134** - Intermediate level technicians from legal, social, sports, cultural and similar services
*   **141** - Office workers, secretaries in general and data processing operators
*   **143** - Data, accounting, statistical, financial services and registry-related operators
*   **144** - Other administrative support staff
*   **151** - personal service workers
*   **152** - sellers
*   **153** - Personal care workers and the like
*   **171** - Skilled construction workers and the like, except electricians
*   **173** - Skilled workers in printing, precision instrument manufacturing, jewelers, artisans and the like
*   **175** - Workers in food processing, woodworking, clothing and other industries and crafts
*   **191** - cleaning workers
*   **192** - Unskilled workers in agriculture, animal production, fisheries and forestry
*   **193** - Unskilled workers in extractive industry, construction, manufacturing and transport
*   **194** - Meal preparation assistants

#### Nationality
*   **1** - Portuguese
*   **2** - German
*   **6** - Spanish
*   **11** - Italian
*   **13** - Dutch
*   **14** - English
*   **17** - Lithuanian
*   **21** - Angolan
*   **22** - Cape Verdean
*   **24** - Guinean
*   **25** - Mozambican
*   **26** - Santomean
*   **32** - Turkish
*   **41** - Brazilian
*   **62** - Romanian
*   **100** - Moldova (Republic of)
*   **101** - Mexican
*   **103** - Ukrainian
*   **105** - Russian
*   **108** - Cuban
*   **109** - Colombian

## 1.1 Import  libraries

Import the libraries needed for data cleaning and analysis

In [14]:
# base libraries
import pandas as pd
import numpy as np


## 1.1.1 Load data

The data for this project was downloaded from UCI Machine Learning Repository. The url is: https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success. The data was downloaded and stored in `data/raw/data.csv`.

In [15]:
df = pd.read_csv('../data/raw/data.csv', sep=';')

df.shape

(4424, 37)

**Interpretation:** The dataset was loaded successfully. This dataset contains 4,424 rows and 37 columns. This indicates that there are `4,424 students` in the dataset and `37 features` for each student.

## 1.1.2 Understand the data

In this step, I will explore the data to understand its structure, identify missing values, and perform basic data analysis. This will help me to identify any issues with the data that need to be addressed before proceeding with data cleaning.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   int64  
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4424 non-null   int64  
 4   Daytime/evening attendance	                     4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

In [17]:
df.head()

,Marital status,Application mode,Application order,Course,Daytime/evening attendance\t,Previous qualification,Previous qualification (grade),Nacionality,Mother's qualification,Father's qualification,...,Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP,Target
0,1,17,5,171,1,1,122.0,1,19,12,...,0,0,0,0,0.000000,0,10.8,1.4,1.74,Dropout
1,1,15,1,9254,1,1,160.0,1,1,3,...,0,6,6,6,13.666667,0,13.9,-0.3,0.79,Graduate
2,1,1,5,9070,1,1,122.0,1,37,37,...,0,6,0,0,0.000000,0,10.8,1.4,1.74,Dropout
3,1,17,2,9773,1,1,122.0,1,38,37,...,0,6,10,5,12.400000,0,9.4,-0.8,-3.12,Graduate
4,2,39,1,8014,0,1,100.0,1,37,38,...,0,6,6,6,13.000000,0,13.9,-0.3,0.79,Graduate


**Interpretation:** The columns datatypes are correct for all columns. However, the column titles are long and contain special characters, making them hard to reuse. I will rename the columns to make them easier to read. I will use the `snake_format`, remove special characters such as quotes and parentheses, and handle long names by shortening them while keeping them descriptive.


## 1.2 Rename Columns

I will update the columns titles using the following table:

|  | Original Column Name | New Column Name |
| :--- | :--- | :--- |
| 1 | Marital status | `marital_status` |
| 2 | Application mode | `application_mode` |
| 3 | Application order | `application_order` |
| 4 | Course | `course` |
| 5 | Daytime/evening attendance | `daytime_evening` |
| 6 | Previous qualification | `prev_qualification` |
| 7 | Previous qualification (grade) | `prev_grade` |
| 8 | Nacionality | `nationality` |
| 9 | Mother's qualification | `mother_qualification` |
| 10 | Father's qualification | `father_qualification` |
| 11 | Mother's occupation | `mother_occupation` |
| 12 | Father's occupation | `father_occupation` |
| 13 | Admission grade | `admission_grade` |
| 14 | Displaced | `displaced` |
| 15 | Educational special needs | `special_needs` |
| 16 | Debtor | `debtor` |
| 17 | Tuition fees up to date | `tuition_fees_up_to_date` |
| 18 | Gender | `gender` |
| 19 | Scholarship holder | `scholarship_holder` |
| 20 | Age at enrollment | `age_at_enrollment` |
| 21 | International | `international` |
| 22 | Curricular units 1st sem (credited) | `cu1_credited` |
| 23 | Curricular units 1st sem (enrolled) | `cu1_enrolled` |
| 24 | Curricular units 1st sem (evaluations) | `cu1_evaluations` |
| 25 | Curricular units 1st sem (approved) | `cu1_approved` |
| 26 | Curricular units 1st sem (grade) | `cu1_grade` |
| 27 | Curricular units 1st sem (without evaluations) | `cu1_no_evaluations` |
| 28 | Curricular units 2nd sem (credited) | `cu2_credited` |
| 29 | Curricular units 2nd sem (enrolled) | `cu2_enrolled` |
| 30 | Curricular units 2nd sem (evaluations) | `cu2_evaluations` |
| 31 | Curricular units 2nd sem (approved) | `cu2_approved` |
| 32 | Curricular units 2nd sem (grade) | `cu2_grade` |
| 33 | Curricular units 2nd sem (without evaluations) | `cu2_no_evaluations` |
| 34 | Unemployment rate | `unemployment_rate` |
| 35 | Inflation rate | `inflation_rate` |
| 36 | GDP | `gdp` |
| 37 | Target | `target` |

The resulting dataset will be saved to `data/interim/` folder

In [18]:
rename_map = {
    'Marital status': 'marital_status',
    'Application mode': 'application_mode',
    'Application order': 'application_order',
    'Course': 'course',
    'Daytime/evening attendance\t': 'daytime_evening',
    'Previous qualification': 'prev_qualification',
    'Previous qualification (grade)': 'prev_grade',
    'Nacionality': 'nationality',
    "Mother's qualification": 'mother_qualification',
    "Father's qualification": 'father_qualification',
    "Mother's occupation": 'mother_occupation',
    "Father's occupation": 'father_occupation',
    'Admission grade': 'admission_grade',
    'Displaced': 'displaced',
    'Educational special needs': 'special_needs',
    'Debtor': 'debtor',
    'Tuition fees up to date': 'tuition_fees_up_to_date',
    'Gender': 'gender',
    'Scholarship holder': 'scholarship_holder',
    'Age at enrollment': 'age_at_enrollment',
    'International': 'international',
    'Curricular units 1st sem (credited)': 'cu1_credited',
    'Curricular units 1st sem (enrolled)': 'cu1_enrolled',
    'Curricular units 1st sem (evaluations)': 'cu1_evaluations',
    'Curricular units 1st sem (approved)': 'cu1_approved',
    'Curricular units 1st sem (grade)': 'cu1_grade',
    'Curricular units 1st sem (without evaluations)': 'cu1_no_evaluations',
    'Curricular units 2nd sem (credited)': 'cu2_credited',
    'Curricular units 2nd sem (enrolled)': 'cu2_enrolled',
    'Curricular units 2nd sem (evaluations)': 'cu2_evaluations',
    'Curricular units 2nd sem (approved)': 'cu2_approved',
    'Curricular units 2nd sem (grade)': 'cu2_grade',
    'Curricular units 2nd sem (without evaluations)': 'cu2_no_evaluations',
    'Unemployment rate': 'unemployment_rate',
    'Inflation rate': 'inflation_rate',
    'GDP': 'gdp',
    'Target': 'target'
}

df = df.rename(columns=rename_map)
df.to_csv('../data/interim/01.2_data_renamed.csv', index=False)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   marital_status           4424 non-null   int64  
 1   application_mode         4424 non-null   int64  
 2   application_order        4424 non-null   int64  
 3   course                   4424 non-null   int64  
 4   daytime_evening          4424 non-null   int64  
 5   prev_qualification       4424 non-null   int64  
 6   prev_grade               4424 non-null   float64
 7   nationality              4424 non-null   int64  
 8   mother_qualification     4424 non-null   int64  
 9   father_qualification     4424 non-null   int64  
 10  mother_occupation        4424 non-null   int64  
 11  father_occupation        4424 non-null   int64  
 12  admission_grade          4424 non-null   float64
 13  displaced                4424 non-null   int64  
 14  special_needs           

**Interpretation:** The columns have been renamed to improve readability and consistency.

**Note**: An interim dataset was created and saved to `/data/interim/01.2.data_renamed.csv` to facilitate further analysis, however this is not the final file as I will continue to refine the dataset for analysis.

## 1.2.1 Validate interim data

Validate that loaded correactly. The dataset the I will be using from now on is `data/interim/01.2_data_renamed.csv`

In [21]:
# load interim dataset that was saved from previous step. Next time I run this notebook. I can start running from this point instead of need to clean the data again from the beginning.
df = pd.read_csv('../data/interim/01.2_data_renamed.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   marital_status           4424 non-null   int64  
 1   application_mode         4424 non-null   int64  
 2   application_order        4424 non-null   int64  
 3   course                   4424 non-null   int64  
 4   daytime_evening          4424 non-null   int64  
 5   prev_qualification       4424 non-null   int64  
 6   prev_grade               4424 non-null   float64
 7   nationality              4424 non-null   int64  
 8   mother_qualification     4424 non-null   int64  
 9   father_qualification     4424 non-null   int64  
 10  mother_occupation        4424 non-null   int64  
 11  father_occupation        4424 non-null   int64  
 12  admission_grade          4424 non-null   float64
 13  displaced                4424 non-null   int64  
 14  special_needs           

**Interpretation:** The da consist of 37 columns. Some columms with datatype `int64` are the binary column which only accept only 0 or 1.

## 1.2.3 Check for NaN values

Validate that there are no Nan values in dataset. If they are, I will handle them appropriately.

In [27]:
df.isna().sum()

marital_status             0
application_mode           0
application_order          0
course                     0
daytime_evening            0
prev_qualification         0
prev_grade                 0
nationality                0
mother_qualification       0
father_qualification       0
mother_occupation          0
father_occupation          0
admission_grade            0
displaced                  0
special_needs              0
debtor                     0
tuition_fees_up_to_date    0
gender                     0
scholarship_holder         0
age_at_enrollment          0
international              0
cu1_credited               0
cu1_enrolled               0
cu1_evaluations            0
cu1_approved               0
cu1_grade                  0
cu1_no_evaluations         0
cu2_credited               0
cu2_enrolled               0
cu2_evaluations            0
cu2_approved               0
cu2_grade                  0
cu2_no_evaluations         0
unemployment_rate          0
inflation_rate

***Interpretation:*** No nulls values found in the dataset. All columns have complete data.

## 1.3 Columns Datatypes Analyze

The dataset contains categorical, numerical, and binary columns. I will analyze the datatypes of each column to ensure they are appropriate for the data and to identify potential memory optimization opportunities.

## 1.3 Columns Analysis

I will analyze the columns to ensure they are appropriate for the data and to identify potential memory optimization opportunities.

In [38]:
# function to print how much time a unique value appears in a column
def print_nunique_values_by_cols(df, cols):
    for col in cols:
        print(f"{col}: {df[col].nunique()} unique values")

# prints all unique values in a column
def print_unique_values_by_cols(df,cols):
    for col in cols:
        print(f"{col}: {df[col].unique()}")


## 1.3.1 Target Column

The target column is a categorial column with three values: `Dropout`, `Graduate`, and `Enrolled`. I will validate for unique values to ensure data integrity and to identify potential issues.

In [36]:
# Check for unique values in categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns
print_nunique_values_by_cols(df, categorical_cols)

target: 3 unique values


In [39]:
# Print unique values for each column
print_unique_values_by_cols(df, categorical_cols)

target: ['Dropout' 'Graduate' 'Enrolled']


***Interpretation:*** From the otput above, I can confirm the target column is a categorial column with three values: `Dropout`, `Graduate`, and `Enrolled`.

## 1.3.2 Numerical columns

In [28]:
# Check for unique values in categorical columns
categorical_cols = df.select_dtypes(include=['int64']).columns
for col in categorical_cols:
    print(f"{col}: {df[col].nunique()} unique values")

marital_status: 6 unique values
application_mode: 18 unique values
application_order: 8 unique values
course: 17 unique values
daytime_evening: 2 unique values
prev_qualification: 17 unique values
nationality: 21 unique values
mother_qualification: 29 unique values
father_qualification: 34 unique values
mother_occupation: 32 unique values
father_occupation: 46 unique values
displaced: 2 unique values
special_needs: 2 unique values
debtor: 2 unique values
tuition_fees_up_to_date: 2 unique values
gender: 2 unique values
scholarship_holder: 2 unique values
age_at_enrollment: 46 unique values
international: 2 unique values
cu1_credited: 21 unique values
cu1_enrolled: 23 unique values
cu1_evaluations: 35 unique values
cu1_approved: 23 unique values
cu1_no_evaluations: 11 unique values
cu2_credited: 19 unique values
cu2_enrolled: 22 unique values
cu2_evaluations: 30 unique values
cu2_approved: 20 unique values
cu2_no_evaluations: 10 unique values


***Interpretation:*** The columns with two unique values are binary columns with values 1/0. This is expected as these columns represent binary attributes. The other columns have more than two unique values which means that they are categorical columns with more than two categories. In this dataset, the categories are already encoded. See Variable Table above.

## 1.3.3 Binary Columns

I will analyze the binary columns description (`df.description()`) and check for any potential issues or inconsistencies.

In [55]:
binary_cols = ['displaced', 'special_needs', 'debtor', 'tuition_fees_up_to_date', 'gender', 'scholarship_holder', 'international', 'daytime_evening']
df[binary_cols].info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   displaced                4424 non-null   int8 
 1   special_needs            4424 non-null   int8 
 2   debtor                   4424 non-null   int8 
 3   tuition_fees_up_to_date  4424 non-null   int8 
 4   gender                   4424 non-null   int8 
 5   scholarship_holder       4424 non-null   int8 
 6   international            4424 non-null   int8 
 7   daytime_evening          4424 non-null   int8 
dtypes: int8(8)
memory usage: 34.7 KB


**Interpretation:** The binary columns have only two possible values, 1 or 0. The count column shows the number of non null values, which is the same as the total number of rows in the dataset, therefore, there are no missing values in binary columns. The datatype is `int64`, this can be optimized to `int8` which use 1 byte per value. I will convert the datatype to `int8` to reduce memory usage.


In [56]:
df[binary_cols] = df[binary_cols].astype('int8')
df[binary_cols].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   displaced                4424 non-null   int8 
 1   special_needs            4424 non-null   int8 
 2   debtor                   4424 non-null   int8 
 3   tuition_fees_up_to_date  4424 non-null   int8 
 4   gender                   4424 non-null   int8 
 5   scholarship_holder       4424 non-null   int8 
 6   international            4424 non-null   int8 
 7   daytime_evening          4424 non-null   int8 
dtypes: int8(8)
memory usage: 34.7 KB


**Interpretation:** All binary columns datatype have been udpated to `int8`.

In [57]:
df[binary_cols].describe()


,displaced,special_needs,debtor,tuition_fees_up_to_date,gender,scholarship_holder,international,daytime_evening
count,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000
mean,0.548373,0.011528,0.113698,0.880651,0.351718,0.248418,0.024864,0.890823
std,0.497711,0.106760,0.317480,0.324235,0.477560,0.432144,0.155729,0.311897
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
50%,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
75%,1.000000,0.000000,0.000000,1.000000,1.000000,0.000000,0.000000,1.000000
max,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000


***Interpretation:*** The binary columns have only two unique values, which is expected as they represent binary attributes. The count column shows the number of non-null values, which is the same as the total number of rows in the dataset. The mean column shows the proportion of 1s in each column, which can be used to understand the prevalence of each binary attribute. The std column shows the standard deviation of the binary values, which can be used to understand the variability of each binary attribute. The min and max columns show the minimum and maximum values, which should be 0 and 1, respectively. The 25%, 50%, and 75% columns show the quartiles of the binary values, which can be used to understand the distribution of each binary attribute.

The columns `['displaced']` have a mean of `0.548` and a standard deviation close to `0.5`, indicating this column is balanced between 0 and 1, with similar numbers of students in each group.

The columns `['gender','scholarship_holder'] have mean of `0.35`, and `0.25`, indicating that they are not perfecly balance but is well spread between 0 and 1.

The column `daytime_evening', with mean `0.89` indicates that `89%` of students attend during the day and only about `11%` attende in the evening.

For the column `debtor` the mean is `0.11`, this means that `11%` are currently debtors which is a minority.


## 1.3.4 Categorical Columns

I will analyze the categorical columns description (`df.description()`) and check for any potential issues or inconsistencies.

In [71]:
categorical_cols = [
    'marital_status',
    'application_mode',
    'application_order',
    'course',
    'prev_qualification',
    'nationality',
    'mother_qualification',
    'father_qualification',
    'mother_occupation',
    'father_occupation'
]

print_nunique_values_by_cols(df,categorical_cols)
print('------')
print(df[categorical_cols].info())


marital_status: 6 unique values
application_mode: 18 unique values
application_order: 8 unique values
course: 17 unique values
prev_qualification: 17 unique values
nationality: 21 unique values
mother_qualification: 29 unique values
father_qualification: 34 unique values
mother_occupation: 32 unique values
father_occupation: 46 unique values
------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   marital_status        4424 non-null   category
 1   application_mode      4424 non-null   category
 2   application_order     4424 non-null   category
 3   course                4424 non-null   category
 4   prev_qualification    4424 non-null   category
 5   nationality           4424 non-null   category
 6   mother_qualification  4424 non-null   category
 7   father_qualification  4424 non-null   category
 8   mother_occupation

**Interpretation:** The categorical columns have a wide range of unique values, which is expected as they represent different attributes. The count column shows the number of non-null values, which is the same as the total number of rows in the dataset. The datatype is `int64` which can be optimized by using dataype `category`. I will update the datatype to use the correct datatype `category`.

In [68]:
df[categorical_cols] = df[categorical_cols].astype('category')
df[categorical_cols].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   marital_status        4424 non-null   category
 1   application_mode      4424 non-null   category
 2   application_order     4424 non-null   category
 3   course                4424 non-null   category
 4   prev_qualification    4424 non-null   category
 5   nationality           4424 non-null   category
 6   mother_qualification  4424 non-null   category
 7   father_qualification  4424 non-null   category
 8   mother_occupation     4424 non-null   category
 9   father_occupation     4424 non-null   category
dtypes: category(10)
memory usage: 51.9 KB


**Interpretation:** All binary columns datatype have been udpated to `category`.


In [70]:
df[categorical_cols].describe()


,marital_status,application_mode,application_order,course,prev_qualification,nationality,mother_qualification,father_qualification,mother_occupation,father_occupation
count,4424,4424,4424,4424,4424,4424,4424,4424,4424,4424
unique,6,18,8,17,17,21,29,34,32,46
top,1,1,1,9500,1,1,1,37,9,9
freq,3919,1708,3026,766,3717,4314,1069,1209,1577,1010


***Interpretation:*** From above result, the top and feq values tell me about the mode, the most common value, and how often it appears. For nationality, the top value is 'Portugal' with a frequency of 1000, indicating that most students are from Portugal, aproximateely `97%` (4314/4424) * 100.

## 1.3.4 Numerical Columns

I will analyze the numerical columns description (`df.description()`) and check for any potential issues or inconsistencies.

In [75]:
numeric_cols = [
    'admission_grade',
    'cu1_credited',
    'cu1_enrolled',
    'cu1_evaluations',
    'cu1_approved',
    'cu1_grade',
    'cu1_no_evaluations',
    'cu2_credited',
    'cu2_enrolled',
    'cu2_evaluations',
    'cu2_approved',
    'cu2_grade',
    'cu2_no_evaluations',
    'age_at_enrollment'
]

df[numeric_cols].describe()

,admission_grade,cu1_credited,cu1_enrolled,cu1_evaluations,cu1_approved,cu1_grade,cu1_no_evaluations,cu2_credited,cu2_enrolled,cu2_evaluations,cu2_approved,cu2_grade,cu2_no_evaluations,age_at_enrollment
count,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000
mean,126.978119,0.709991,6.270570,8.299051,4.706600,10.640822,0.137658,0.541817,6.232143,8.063291,4.435805,10.230206,0.150316,23.265145
std,14.482001,2.360507,2.480178,4.179106,3.094238,4.843663,0.690880,1.918546,2.195951,3.947951,3.014764,5.210808,0.753774,7.587816
min,95.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,17.000000
25%,117.900000,0.000000,5.000000,6.000000,3.000000,11.000000,0.000000,0.000000,5.000000,6.000000,2.000000,10.750000,0.000000,19.000000
50%,126.100000,0.000000,6.000000,8.000000,5.000000,12.285714,0.000000,0.000000,6.000000,8.000000,5.000000,12.200000,0.000000,20.000000
75%,134.800000,0.000000,7.000000,10.000000,6.000000,13.400000,0.000000,0.000000,7.000000,10.000000,6.000000,13.333333,0.000000,25.000000
max,190.000000,20.000000,26.000000,45.000000,26.000000,18.875000,12.000000,19.000000,23.000000,33.000000,20.000000,18.571429,12.000000,70.000000


**Interpretation:** The numerical columns have a wide range of values, which is expected as they represent different attributes. The min column shows the minimum value of each column, which can be used to understand the minimum value of each numerical attribute. The max column shows the maximum value of each column, which can be used to understand the maximum value of each numerical attribute.

The `age_at_enrollment` column shows that most students enter around age 19–20, with a minimum value of 17 and maximum of 70 which is fine for higher education. Also the `75%` shows the most common age range for enrollment is 20-25. From this, I can see that it may be appropiate to create a categories for age called `age_group` which could be `['17-19','20-25','25-30', '30-70']`.



## Data preparation or Feature Engineering

## Analysis of Results and Conclusions

Description of the results obtained and if there are conclusions that can be drawn from them.

The analysis of results must be related to the description of the task.

**Note:** An analysis of results does not necessarily lead to conclusions, but to ideas or proposals for future work


## Proposals and Ideas

From the results obtained, what ideas or proposals can be generated to continue with the project
